File: Hard/longest_valid_parentheses.py
class Solution(object):
    def longestValidParentheses(self, s):
        """
        :type s: str
        :rtype: int
        """
        countValidParentheses = 0
        storeIndex = [-1]

        for i, c in enumerate(s):
            if c == "(":
                storeIndex.append(i)
            else:
                storeIndex.pop()
        
                if len(storeIndex) == 0:
                    storeIndex.append(i)
                else:
                    countValidParentheses = max(countValidParentheses, i - storeIndex[-1])
                
        return countValidParentheses
    
print(Solution().longestValidParentheses("(()")) #2
print(Solution().longestValidParentheses(")()())")) #4
print(Solution().longestValidParentheses("")) #0
print(Solution().longestValidParentheses("()(()")) #2
print(Solution().longestValidParentheses("()(())")) #6
print(Solution().longestValidParentheses("()")) #6

File: Hard/longest_valid_parentheses.js
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function (s) {
  let countValidParentheses = 0;
  let storeIndex = [-1];

  for (let i = 0; i < s.length; i++) {
    if (s[i] == "(") {
      storeIndex.push(i);
    } else {
      storeIndex.pop();
      if (storeIndex.length == 0) {
        storeIndex.push(i);
      } else {
        countValidParentheses = Math.max(countValidParentheses, i - storeIndex[storeIndex.length - 1]);
      }
    }
  }

  return countValidParentheses;
};

console.log(longestValidParentheses("(()")); //2
console.log(longestValidParentheses(")()())")); //4
console.log(longestValidParentheses("")); //0
console.log(longestValidParentheses("()(()")); //2
console.log(longestValidParentheses("()(())")); //6
console.log(longestValidParentheses("()")); //6
